dev/test/bulk-writer.ts

```diff
@@ -0,0 +1,457 @@
+    const responses = arrayRange.map(i => successResponse(i));
+    await instantiateInstance([
+      {
+        request: createRequest([requests[0], requests[1]]),
+        response: createResponse([responses[0], responses[1]]),
+      },
+      {
+        request: createRequest([requests[2], requests[3]]),
+        response: createResponse([responses[2], responses[3]]),
+      },
+      {
+        request: createRequest([requests[4], requests[5]]),
+        response: createResponse([responses[4], responses[5]]),
+      },
+    ]);
+
+    bulkWriter.setMaxBatchSize(2);
+    for (let i = 0; i < 6; i++) {
+      bulkWriter
+        .set(firestore.doc('col/doc' + i), {foo: 'bar'})
+        .then(incrementOpCount);
+    }
+
+    return bulkWriter.flush().then(async () => {
+      verifyOpCount(6);
+    });
+  });
+
+  it('throttles writes', async () => {
+    await instantiateInstance(
+      [
+        {
+          request: createRequest([
+            setOp('col/doc1', 'bar'),
+            setOp('col/doc2', 'bar'),
+          ]),
+          response: createResponse([successResponse(1), successResponse(2)]),
+        },
+        {
+          request: createRequest([setOp('col/doc3', 'bar')]),
+          response: createResponse([successResponse(3)]),
+        },
+      ],
+      /** manualFlush= */ true
+    );
+
+    // TODO: figure out how to properly create barrier to check for throttling
+    // in order to test for requests that are in flight.
+    // GOAL: flush1 runs, 1st batch commits, flush2 runs
+    // ACTUAL: flush1 runs, flush 2 runs, 1st batch commits
+    bulkWriter.setMaxOpsPerSecond(2);
+    bulkWriter.setMaxBatchSize(2);
+    bulkWriter.set(firestore.doc('col/doc1'), {foo: 'bar'});
+    bulkWriter.set(firestore.doc('col/doc2'), {foo: 'bar'});
+    const flush1 = bulkWriter.flush();
+    // The third write will be added to a new batch that should not be allowed
+    // in the same request because it would exceed the rate limit.
+    bulkWriter.set(firestore.doc('col/doc3'), {foo: 'bar'});
+    const flush2 = bulkWriter.flush();
+    flushDeferred.resolve();
+  });
+
+  it('does not add writes to batches that are already in flight', () => {});
+
+  describe('555 support', () => {
+    let observedMax: number[] = [];
+
+    before(() => {
+      setTimeoutHandler((callback, timeout) => {
+        observedMax.push(timeout);
+        callback();
+      });
+    });
+
+    beforeEach(() => {
+      observedMax = [];
+    });
+
+    afterEach(() => setTimeoutHandler(setTimeout));
+
+    it('gradually ramps up the maximum operations per second', () => {
+      const nop = () => {};
+      bulkWriter.set(firestore.doc('col/doc'), {});
+
+      // TODO: Should I finish the test and add hooks to monitor that the maxOps
+      // increases by 50% each time? It seems like a quite a bit of boilerplate to test
+      // something pretty simple. On the other hand, it feels bad to not test it at all.
```

Comment:
Yes you should:) I think getting this right will be quite tricky and tests will help.